### PR TITLE
DTGB-684: Added file upload component

### PR DESCRIPTION
**Pull request by hstandaert**
## Description

* Added new component for file upload
* Other form component optimisations:
	* Make field message colors themeable
	* Add class for form steps list instead of binding it to the element

@mattiasberlamont, you could add the documentation for File Upload in this PR if you have the time. 😉

## Screenshots (if appropriate):
Screenshot 2019-04-23 at 10 32 21

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have updated the documentation accordingly.
- [x] I have updated the styleguide CHANGELOG accordingly.
- [x] I have ran gulp axe on the compiled files and found no errors.

### CODE REVIEW

File: components/00-settings/_colors.scss [hunk 1]
```diff
@@ -205,6 +205,8 @@ $themes: (
     'color-primary': color('cyan'),
     'color-secondary': color('white'),
     'color-tertiary': color('dark-gray'),
+    'color-error': color('red'),
```
Comment (delrueba) [inline components/00-settings/_colors.scss]: These were not themeable by design to implement consistency over all themes and prevent 'artistic' interpretations of error and success colors.

File: components/31-molecules/file-upload/file-upload.functions.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Update the filename of the uploaded file.
+     *
+     * @param {Event} e
+     *  Event object for the change event.
+     *
+     */
+    const updateFile = e => {
+      if (!input.files.length) {
+        elem.dataset.file = options.emptyText || elem.dataset.emptyText || 'No file chosen.';
```
Comment (delrueba) [inline components/31-molecules/file-upload/file-upload.functions.js]: nice use of dataset! :trophy:

File: components/31-molecules/form-steps/_form-steps.scss [hunk 3]
```diff
@@ -1,6 +1,6 @@
 .form-steps,
 .wizard-steps { /// @deprecated
-  & ol {
+  .form-steps-list {
```
Comment (delrueba) [inline components/31-molecules/form-steps/_form-steps.scss]:
```suggestion
  ol.form-steps-list {
```
I would like to enforce developers to use 'ol'.

File: components/31-molecules/form-steps/_form-steps.scss [hunk 4]
```diff
@@ -151,7 +151,7 @@
     }
   }
 
-  &:not(.vertical) ol {
+  &:not(.vertical) .form-steps-list {
```
Comment (delrueba) [inline components/31-molecules/form-steps/_form-steps.scss]:
```suggestion
  &:not(.vertical) ol.form-steps-list {
```

File: components/31-molecules/form-steps/_form-steps.scss [hunk 5]
```diff
@@ -166,7 +166,7 @@
     }
   }
 
-  &.vertical ol {
+  &.vertical .form-steps-list {
```
Comment (delrueba) [inline components/31-molecules/form-steps/_form-steps.scss]:
```suggestion
  &.vertical ol.form-steps-list {
```

**Review by delrueba — CHANGES_REQUESTED** [review]

**Comment by mattiasberlamont** [comment]
@hstandaert Documentation for the File Upload component was added :) Can be still expanded when necessary.

**Comment by mattiasberlamont** [comment]
Is (an example of) the file upload component also worked out for the case where multiple files can be uploaded?

image

**Comment by delrueba** [comment]
> Is (an example of) the file upload component also worked out for the case where multiple files can be uploaded?
> 
> image

About that, i'm looking at this for LEZ right now and there is no option to remove attachments.

**Comment by mattiasberlamont** [comment]
> > Is (an example of) the file upload component also worked out for the case where multiple files can be uploaded?
> > image
> 
> About that, i'm looking at this for LEZ right now and there is no option to remove attachments.

Added an alternative representation for the file upload with an option to remove attachments. As discussed, to be tried out for LEZ.

image

**Review by delrueba — APPROVED** [review]

**Review by delrueba — CHANGES_REQUESTED** [review]
Missing focus style!

File: components/31-molecules/file-upload/file-upload.functions.js [hunk 6]
```diff
@@ -44,12 +44,21 @@
       elem.dataset.file = input.files[0].name;
     };
 
+    /**
+     * Make container element focusable instead of the file input.
+     */
+    const initFocus = () => {
```
Comment (delrueba) [inline components/31-molecules/file-upload/file-upload.functions.js]: It's best not to change this.
It'll break keyboard functionality and screenreaders can miss vital information that's embedded in the input element.

I suggest using :focus-within in combination with some js magic for Edge support.
Or changing the template so we can use the sibling selector.

**Review by delrueba — CHANGES_REQUESTED** [review]

**Comment by hstandaert** [comment]
I included some variants now:

> **NOTE:** These variants are not functional and implementation should be done in the project using it.

localhost_3000_components_preview_file-upload

File: components/31-molecules/file-upload/file-upload.functions.js (hunk 6 above)
Comment (hstandaert) [reply components/31-molecules/file-upload/file-upload.functions.js]: Template has been changed 🙃

File: components/21-atoms/file/file.twig [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<div class="file">
```
Comment (delrueba) [inline components/21-atoms/file/file.twig]:
```suggestion
<div class="file" data-file="No file chosen.">
```

**Review by delrueba — COMMENTED** [review]
Suggest adding data-file to template, see git suggestion.
Otherwise: :+1:

File: components/21-atoms/file/_file.scss [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    &:disabled {
+      pointer-events: none;
+
+      + .file__button {
+        @include button-disabled;
+      }
+    }
+
+    &:hover + .file__button,
+    &:focus + .file__button {
```
Comment (delrueba) [inline components/21-atoms/file/_file.scss]:
```suggestion
    &:focus + .file__button,
    &:focus-within + .file__button {
```

**Review by delrueba — APPROVED** [review]

**Pull request merged**